src/glang/lib/std/rb_tree.c3

```diff
@@ -0,0 +1,481 @@
+        let u : RB_Node<K>& = &u_ptr:data():data();
+
+        if key == u.key {
+            // Key already exists. Overwrite it with the new key (if the tree is
+            // used to implement a map, then the associated value could be
+            // different) and return early, as the red-black tree is already
+            // valid.
+            u.key = key;
+            return;
+        }
+
+        // Keep track of the parent node.
+        p_opt:store(&u_ptr:data():data());
+
+        u_ptr = {&u:get_child(left_if(key < u.key))};
+    }
+
+    u_ptr:data():store(&self:new<RB_Node<K>>(key, p_opt));
+
+    if p_opt:has_value() {
+        let p : RB_Node<K>& = &p_opt:data();
+        p:get_child(left_if(key < p.key)) = u_ptr:data();
+    }
+
+    self:fix_after_insert(&u_ptr:data():data());
+}
+
+function [K] search_impl : (
+    self : RB_Tree<K>&, key : K
+) -> Optional<RB_Node<K>&> = {
+    let u : Optional<RB_Node<K>&> = self.root;
+
+    while u:has_value() and key != u:data().key {
+        u = u:data():get_child(left_if(key < u:data().key));
+    }
+
+    return u;
+}
+
+function [K] search : (self : RB_Tree<K>&, key : K) -> Optional<K> = {
+    let result : Optional<RB_Node<K>&> = self:search_impl(key);
+
+    if result:has_value() {
+        return make_optional(result:data().key);
+    }
+
+    return make_optional<K>();
+}
+
+function [K] shift_nodes : (
+    self : RB_Tree<K>&, old : RB_Node<K>&, new : Optional<RB_Node<K>&>
+) -> void = {
+    // Make old's parent point to new instead of old.
+    if old.parent:has_value() {
+        old.parent:data():get_child(old:which_child_am_i()) = new;
+    }
+    else {
+        self.root = new;
+    }
+
+    // Set new's new parent, if it's not a leaf node (nil node).
+    if new:has_value() {
+        new:data().parent = old.parent;
+    }
+}
+
+function [K] minmax : (node : RB_Node<K>&, direction : bool) -> RB_Node<K>& = {
+    let u : Ptr<RB_Node<K>> = {&node};
+
+    // TODO some sort of lambda/local function would really help here.
+    while u:data():get_child(direction):has_value() {
+        u = {&u:data():get_child(direction):data()};
+    }
+
+    return &u:data();
+}
+
+function [K] minmax : (tree : RB_Tree<K>&, direction : bool) -> Optional<K> = {
+    if tree.root:has_value() {
+        return make_optional(tree.root:data():minmax(direction).key);
+    }
+    return make_optional<K>();
+}
+
+function [K] minimum : (tree : RB_Tree<K>&) -> Optional<K> = {
+    return tree:minmax(LEFT());
+}
+
+function [K] maximum : (tree : RB_Tree<K>&) -> Optional<K> = {
+    return tree:minmax(RIGHT());
+}
+
+function [K] fix_after_delete_impl : (
+    self : RB_Tree<K>&, x : Optional<RB_Node<K>&>, y : RB_Node<K>&
+) -> Optional<RB_Node<K>&> = {
+    // x is guaranteed not to be the root. y is guaranteed to be x's parent.
+    let x_side : bool = which_child_am_i(x, &y);
+    let w_side : bool = !x_side;
+
+    // x's sibling, w, cannot be a leaf node (nil node). If x is double black,
```

Comment:
~~Why can't both `x` and `w` be nil leaf nodes if `x` is "black-and-red"?~~
Ohh, leaf nodes can only ever be double black right?

Reply:
Yeah, the implicit assumption in all of this is that you start with a valid red-black tree, so leaf nodes are always (originally) black :)